### PR TITLE
PERF: Enable %z in parsing datetime

### LINKED ISSUE #32792: to_datetime() throws ValueError: Cannot pass a tz argument when parsing strings with timezone information.
I think pandas should support passing `%z` in the format but also `utc=True`. In my opinion, one thing is the format, which tells pandas how to parse the datetime string. The other argument is just telling to return the dates in UTC, no matter which timezone they were in the beginning.

Here's a repl that shows the issue: https://repl.it/@eparizzi/Pandas-todatetime-in-UTC-with-format

If you replace that simple CSV with some big 50K row time-series CSV, the call to `to_datetime` without the format takes more than 20 seconds. On the contrary, passing the format and without `utc=True` takes less than 2 seconds. Unfortunately, this doesn't seem to work properly when there are multiple timezones in the column. It simply can't set a proper dtype in this case.

So, why can't we have a way to specify the format including timezone but also specify that we want everything in datetime64(UTC)?

I've already gone over this issue: https://github.com/pandas-dev/pandas/issues/25571 but I still think this deserves a discussion.

```python
import pandas as pd

# I know the format, I want to use it so that Pandas to_datetime() runs faster.
DATETIME_FORMAT = '%m/%d/%Y %H:%M:%S.%f%z'

try:
  data = ['10/11/2018 00:00:00.045-07:00',
  '10/11/2018 01:00:00.045-07:00',
  '10/11/2018 01:00:00.045-08:00',
  '10/11/2018 02:00:00.045-08:00',
  '10/11/2018 04:00:00.045-07:00',
  '10/11/2018 05:00:00.045-07:00']

  df = pd.DataFrame(data, columns=["Timestamp"])

  # This raises "ValueError: Cannot pass a tz argument when parsing strings with timezone information."
  df.Timestamp = pd.to_datetime(df.Timestamp, format=DATETIME_FORMAT, utc=True)

except ValueError as valueError:
  # I don't know why a %z in the format is not compatible with utc=True. The %z is telling pandas that it needs to deal with timezones. Then, utc=True should just convert all to UTC. It shouldn't be more complicated than that I think.
  print(f"ERROR: {str(valueError)}")
  print("...why not?")

  # This works, but it's A LOT slower when parsing a lot of rows.
  df.Timestamp = pd.to_datetime(df.Timestamp, infer_datetime_format=True, utc=True)

finally:
  print(df)
  print(df.Timestamp.dtype)

  # Expected output:
  # 
  #                          Timestamp
  # 0 2018-10-11 07:00:00.045000+00:00
  # 1 2018-10-11 08:00:00.045000+00:00
  # 2 2018-10-11 09:00:00.045000+00:00
  # 3 2018-10-11 10:00:00.045000+00:00
  # 4 2018-10-11 11:00:00.045000+00:00
  # 5 2018-10-11 12:00:00.045000+00:00
  # datetime64[ns, UTC]
```

### PR DESCRIPTION
**Pull request by quangngd**
- [ ] closes #32792
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/tools/datetimes.py [hunk 1]
```diff
@@ -229,13 +229,12 @@ def _return_parsed_timezone_results(result, timezones, tz, name):
...
-    if tz is not None:
-        raise ValueError(
-            "Cannot pass a tz argument when parsing strings with timezone information."
-        )
     tz_results = np.array(
         [Timestamp(res).tz_localize(zone) for res, zone in zip(result, timezones)]
     )
+    if tz is not None:
+        # GH 32792
+        tz_results = np.array([tz_result.astimezone(tz) for tz_result in tz_results])
```
Comment (mroeschke) [inline pandas/core/tools/datetimes.py]: Could you use `tz_convert` here instead of `astimezone`?

File: pandas/tests/tools/test_to_datetime.py [hunk 2]
```diff
@@ -323,8 +323,9 @@ def test_to_datetime_parse_tzname_or_tzoffset(self, fmt, dates, expected_dates):
         expected = pd.Index(expected_dates)
         tm.assert_equal(result, expected)
 
-        with pytest.raises(ValueError):
-            pd.to_datetime(dates, format=fmt, utc=True)
+        # GH 32792
+        result = pd.to_datetime(dates, format=fmt, utc=True)
+        assert result.dtype._tz is pytz.UTC
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_datetime.py]: Could you construct the `expected` index and use `tm.assert_index.equal`?

```
result = pd.to_datetime(dates, format=fmt, utc=True)
expected = DatetimeIndex(...)
tm.assert_index_equal(result, expected)
```

**Review by mroeschke — CHANGES_REQUESTED** [review]
Could you add an ASV benchmark for this?

Could you add a whatsnew note for this change (1.1)?

File: pandas/core/tools/datetimes.py (hunk 1 above)
Comment (quangngd) [reply pandas/core/tools/datetimes.py]: @mroeschke sorry but i'm not aware of any ```tz_convert``` which is for a pd.Index which object dtype

**Comment by quangngd** [comment]
Is this really a performance issue? It feels like an enh to free up a use case that previously have to use workaround. I feel weird putting it under ```Performance improvements``` in whatsnew.

File: pandas/core/tools/datetimes.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/tools/datetimes.py]: `tz_convert` is a `Timestamp` method.

**Comment by mroeschke** [comment]
Sure you can put in in the enhancement section

File: pandas/tests/tools/test_to_datetime.py [hunk 3]
```diff
@@ -323,8 +323,11 @@ def test_to_datetime_parse_tzname_or_tzoffset(self, fmt, dates, expected_dates):
         expected = pd.Index(expected_dates)
         tm.assert_equal(result, expected)
 
-        with pytest.raises(ValueError):
-            pd.to_datetime(dates, format=fmt, utc=True)
+        # GH 32792
```
Comment (jreback) [inline pandas/tests/tools/test_to_datetime.py]: make a new test here

File: doc/source/whatsnew/v1.1.0.rst [hunk 4]
```diff
@@ -70,6 +70,7 @@ Other enhancements
 - :func:`timedelta_range` will now infer a frequency when passed ``start``, ``stop``, and ``periods`` (:issue:`32377`)
 - Positional slicing on a :class:`IntervalIndex` now supports slices with ``step > 1`` (:issue:`31658`)
 - :meth:`DataFrame.sample` will now also allow array-like and BitGenerator objects to be passed to ``random_state`` as seeds (:issue:`32503`)
+- :func:`to_datetime` will now allow parsing formats with ``%z`` while ``utc=True`` (:issue:`32792`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: expand this note to talk about time zone strings as %z isn’t obvious to all ; even show a partial format

File: asv_bench/benchmarks/timeseries.py [hunk 5]
```diff
@@ -336,15 +336,33 @@ def time_infer_quarter(self):
...
         to_datetime(self.s2, format="%d%b%y")
 
     def time_no_exact(self):
         to_datetime(self.s, format="%d%b%y", exact=False)
 
+    def time_same_offset(self):
+        to_datetime(self.same_offset, format="%m/%d/%Y %H:%M:%S.%f%z")
+
+    def time_different_offset(self):
+        to_datetime(self.same_offset, format="%m/%d/%Y %H:%M:%S.%f%z")
```
Comment (jreback) [inline asv_bench/benchmarks/timeseries.py]: shouldn’t this be diff_offset

File: pandas/core/tools/datetimes.py [hunk 6]
```diff
@@ -229,13 +229,12 @@ def _return_parsed_timezone_results(result, timezones, tz, name):
...
     tz_result : Index-like of parsed dates with timezone
     """
-    if tz is not None:
-        raise ValueError(
-            "Cannot pass a tz argument when parsing strings with timezone information."
-        )
     tz_results = np.array(
         [Timestamp(res).tz_localize(zone) for res, zone in zip(result, timezones)]
     )
+    if tz is not None:
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: the issue number is not important instead add a comment in what / why u r doing this here

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.1.0.rst [hunk 7]
```diff
@@ -58,6 +58,22 @@ For example:
 
 For more on working with fold, see :ref:`Fold subsection <timeseries.fold>` in the user guide.
 
+.. _whatsnew_110.to_datetime_multiple_tzname_tzoffset_support:
+
+Parsing timezone-aware format with different timezones in to_datetime
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:func:`to_datetime` now supports parsing formats containing timezone names (``%Z``) and UTC offsets (``%z``) from different timezones then converting them to UTC by setting ``utc=True``. This would returns a ``DatetimeIndex`` with timezone at UTC as opposed to an ``Index`` with ``object`` dtype if ``utc=True`` is not set (:issue:`32792`).
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.1.0.rst]: ``` would return a :class:`DatetimeIndex` with timezone at UTC as opposed to an :class:`Index` ```

File: pandas/tests/tools/test_to_datetime.py [hunk 8]
```diff
@@ -323,8 +323,26 @@ def test_to_datetime_parse_tzname_or_tzoffset(self, fmt, dates, expected_dates):
...
+            "2010-01-01 11:00:00+00:00",
+            "2010-01-01 13:00:00+00:00",
+            "2010-01-01 09:00:00+00:00",
+            "2010-01-01 08:00:00+00:00",
+        ]
+        fmt = "%Y-%m-%d %H:%M:%S %z"
+
+        result = pd.to_datetime(dates, format=fmt, utc=True)
+        expected = pd.DatetimeIndex(expected_dates)
+        tm.assert_equal(result, expected)
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_datetime.py]: Can you use `tm.assert_equal_index` here?

File: pandas/tests/tools/test_to_datetime.py [hunk 9]
```diff
@@ -323,8 +323,26 @@ def test_to_datetime_parse_tzname_or_tzoffset(self, fmt, dates, expected_dates):
...
+        assert result.dtype._tz is pytz.UTC
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_datetime.py]: Don't need this last `assert` here

**Review by mroeschke — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
very nice @quangngd keep em coming!

**Comment by eparizzi** [comment]
This is amazing. Thank you!

In which release is this going to be available? 1.1 on August?

**Comment by jreback** [comment]
likely 1.1 will be soonish - next month or 2

**Comment by eparizzi** [comment]
Here's a test case notebook showing why pandas should support this:

https://www.kaggle.com/eparizzi/pandas-to-datetime-is-slow

**Comment by jreback** [comment]
@eparizzi this is already fixed and will be released soon

**Pull request merged**